### PR TITLE
chore(build): migrate to `pyproject.toml` for packaging

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
Mirrors work done in:
* https://github.com/google-deepmind/optax/pull/513
* https://github.com/google-deepmind/optax/pull/737

`pyproject.toml` is the new standard for packaging python packages, `setup.py` is now deprecated (first introduced in [PEP 518](https://peps.python.org/pep-0518/) and later expanded in [PEP 517](https://peps.python.org/pep-0517/), [PEP 621](https://peps.python.org/pep-0621/) and [PEP 660](https://peps.python.org/pep-0660/)). This PR proposes to switch from the legacy `setup.py` method in favour of `pyproject.toml`. This would lead to a overall better project structure because all requirements would be managed within a single `pyproject.toml` file instead of 4 separate text files. It also allows for dynamic version allocation (reads automatically from `/__init__.py`). 

Changes proposed by this PR can be summarized as follows :-

* Delete `setup.py`  in favour of `pyproject.toml` for packaging and update the `test.sh` script appropriately.
* Update Github Actions Workflows with `cache` and `cache-dependency-path` and different build commands for uploading to pypi.
* Removing the now redundant `requirements*.txt` and `MANIFEST.in`.
* Updating "read the docs" configuration.

### CODE REVIEW

**Comment by claudiofantacci** [comment]
Thanks for this PR! I'll try to have a look as soon as I can. In the meantime, can you have a look the CI that seems to be failing? Thanks again 🚀

**Comment by SauravMaheshkar** [comment]
> Thanks for this PR! I'll try to have a look as soon as I can. In the meantime, can you have a look the CI that seems to be failing? Thanks again 🚀

Addressed in [495b8a8](https://github.com/google-deepmind/dm_pix/pull/97/commits/495b8a80624f4da429fb1777d5bbdc9e6099c606), it was a typo 😅

**Comment by claudiofantacci** [comment]
Ok seems like that fixed it! 🚀 
I'll have a look asap, but skimming through it LGTM 👍

**Comment by SauravMaheshkar** [comment]
Also maybe it's worth mentioning but with `optax` copybara was throwing some errors and then ultimately someone had to add a empty `pyproject.toml`. If that is the case I can rebase the PR with a single squash commit to make everything nice and clean.

**Comment by claudiofantacci** [comment]
That's strange, in theory it should be a single squashed commit anyway.
In any case, given your prior with `optax` I would suggest doing that.
Can you please do it 😄 ?

**Comment by SauravMaheshkar** [comment]
Happy to do it if necessary. Can you try running capybara on this PR?

Previously the GitHub CI had ran successfully however when attempting to merge, capybara was throwing an error. Can you try to run capybara on this PR? Incase there is an error I'll open another PR with an empty pyproject

File: test.sh [hunk 1]
```diff
@@ -26,20 +26,20 @@ python --version
 
 # Install JAX.
 python -m pip install --upgrade pip setuptools
-python -m pip install -r requirements.txt
-python -c 'import jax; print(jax.__version__)'
 
 # Run setup.py, this installs the python dependencies
```
Comment (claudiofantacci) [inline test.sh:30]: This is no longer what happens, can you please update the comment here?

File: test.sh [hunk 2]
```diff
@@ -26,20 +26,20 @@ python --version
 
 # Install JAX.
```
Comment (claudiofantacci) [inline test.sh:27]: This is also no longer the case, the same thing for line 22. Can you please update the comments to reflect the changed introduced by this PR?

File: test.sh [hunk 3]
```diff
@@ -26,20 +26,20 @@ python --version
...
+python -c 'import jax; print(jax.__version__)'
+
 # Python test dependencies.
-python -m pip install -r requirements_tests.txt
+python -m pip install -e ".[test]"
 
 # CPU count on macos or linux
 if [ "$(uname)" == "Darwin" ]; then
-  N_JOBS=$(sysctl -n hw.logicalcpu)
+	N_JOBS=$(sysctl -n hw.logicalcpu)
```
Comment (claudiofantacci) [inline test.sh:40]: Please leave the two spaces.

File: test.sh [hunk 4]
```diff
@@ -26,20 +26,20 @@ python --version
...
 else
-  N_JOBS=$(grep -c ^processor /proc/cpuinfo)
+	N_JOBS=$(grep -c ^processor /proc/cpuinfo)
```
Comment (claudiofantacci) [inline test.sh:42]: Please leave the two spaces.

**Review by claudiofantacci — COMMENTED** [review]
Thanks for the PR, only a few minor things before approving 😄

**Comment by SauravMaheshkar** [comment]
I addressed the comments in the recent commits.

**Review by claudiofantacci — APPROVED** [review]
Thanks a lot, let's see what Copybara tells us ahah 😄

**Comment by claudiofantacci** [comment]
I have a copybara failure, I'll look into it asap.

**Comment by SauravMaheshkar** [comment]
> I have a copybara failure, I'll look into it asap.

Hey @claudiofantacci, I'm assuming this is because of a lack of `pyproject.toml` in the master branch atm. Can you make a PR via copybara which adds a empty `pyproject.toml` ? Then I can rebase my PR and we can try again.

**Comment by claudiofantacci** [comment]
My Copybara is probably setup differently 😄 
The error I get is because we're deleting some files (the `requitements`) that are expected.
I should be able to change this behaviour quite easily, but I'm juggling through a lot of things so this might require me a bit of time. I apologise! I'll get back here as soon as I can!

**Comment by claudiofantacci** [comment]
Ok, I've prepared internal changes that we're testing to get this PR through 🚀 

In the meantime, can you please update `README.md` according to the changes of this PR?

**Comment by claudiofantacci** [comment]
Ok the PR is in, can you rebase and update the `README.md`? 😄

**Review by claudiofantacci — COMMENTED** [review]
Ok the #98 is in, can you rebase and update the `README.md`?

**Review by claudiofantacci — CHANGES_REQUESTED** [review]
Ok the #98 is in, can you rebase and update the `README.md`?

**Comment by SauravMaheshkar** [comment]
Sorry for the delay, I updated the `README` in [2a2d526](https://github.com/google-deepmind/dm_pix/pull/97/commits/2a2d52689593bd934db30ad5137242cc58f0cb5c)

**Review by claudiofantacci — CHANGES_REQUESTED** [review]

**Review by claudiofantacci — APPROVED** [review]
Amazing, thanks!

**Pull request merged**